pact/constants.py

```diff
@@ -1,38 +1,60 @@
-"""Constant values for the pact-python package."""
+"""
+Constant values for the pact-python package.
+
+This will default to the bundled Pact binaries bundled with the package, but
+should these be unavailable or the environment variable `USE_SYSTEM_PACT` is
+set to `TRUE` or `YES`, the system Pact binaries will be used instead.
+"""
 import os
+import shutil
+import warnings
 from pathlib import Path
 
-
-def broker_client_exe() -> str:
-    """Get the appropriate executable name for this platform."""
-    if os.name == "nt":
-        return "pact-broker.bat"
-    return "pact-broker"
-
-
-def message_exe() -> str:
-    """Get the appropriate executable name for this platform."""
-    if os.name == "nt":
-        return "pact-message.bat"
-    return "pact-message"
-
-
-def mock_service_exe() -> str:
-    """Get the appropriate executable name for this platform."""
-    if os.name == "nt":
-        return "pact-mock-service.bat"
-    return "pact-mock-service"
-
-
-def provider_verifier_exe() -> str:
-    """Get the appropriate provider executable name for this platform."""
-    if os.name == "nt":
-        return "pact-provider-verifier.bat"
-    return "pact-provider-verifier"
-
-
-ROOT_DIR = Path(__file__).parent.resolve()
-BROKER_CLIENT_PATH = ROOT_DIR / "bin" / broker_client_exe()
-MESSAGE_PATH = ROOT_DIR / "bin" / message_exe()
-MOCK_SERVICE_PATH = ROOT_DIR / "bin" / mock_service_exe()
-VERIFIER_PATH = ROOT_DIR / "bin" / provider_verifier_exe()
+__all__ = [
+    "BROKER_CLIENT_PATH",
+    "MESSAGE_PATH",
+    "MOCK_SERVICE_PATH",
+    "VERIFIER_PATH",
+]
+
+
+_USE_SYSTEM_PACT = os.getenv("USE_SYSTEM_PACT", "").upper() in ("TRUE", "YES")
+_BIN_DIR = Path(__file__).parent.resolve() / "bin"
+
+
+def _find_executable(executable: str) -> str:
+    """
+    Find the path to an executable.
+
+    This inspects the environment variable `USE_SYSTEM_PACT` to determine
```

Comment:
does this have to be available on the path?

could the user have the possibly to provider their own path, if they don't want to or struggle to have it available on their path?

I think a `PACT_XYZ` env var might be more idiomatic with env vars set in other languages, the pact broker and the pact cli 

say PACT_USE_SYSTEM_LIBS

you could also then have potentially a CUSTOM_LIBS location.

I think @mefellows would love in a config living in a users `$HOME/.pact` directory and the plugins are now installed into their.

They take a PACT_PLUGIN_DIR location.

As this is for the CLI tools? it could be PACT_CLI_DIR

Reply:
Very good points!

- I will rename the system variable to be `PACT_...`

- The `USE_SYSTEM_PACT` described above was more specifically for the Ruby executables. Upon reflection, I really should have:
  - `PACT_USE_SYSTEM_BINS` for the CLI tools
  - `PACT_USE_SYSTEM_LIBS` for the libraries. Though I'm not entirely sure how/if that will work.

- Specifically regarding `PACT_USE_SYSTEM_LIBS`, at the moment the build script pulls down the static libraries and then creates shared library against the system libraries. Specifically, library created contains the Pact lib in its entirety (and therefore doesn't rely on a `libpact_ffi` being installed in `/usr/lib`).

- The above variables are simply booleans and use the system's default resolution for executables and libraries. I generally don't really like hijacking the way specifying direct paths, as there is a strong Unix convention already of updating `PATH`. For example:

  ```sh
  PACT_USE_SYSTEM_BINS=True PATH=/path/to/pact:$PATH pytest ...
  ```

  This is particularly important, in my opinion, for the Pact libraries too as the loading of libraries is done much less explicitly than the calling of executables. By setting `LD_LIBRARY_PATH` as appropriate, there is much less risk of accidentally loading the wrong library.

- Regarding the Pact plugin's, at the moment, this is not supported at all, but I agree it will need to be done eventually. I don't think it is within the scope of this PR for now and I'll get to it once I implement plugins in the library.